py/desitarget/mtl.py

```diff
@@ -557,20 +593,43 @@ def make_ledger_in_hp(targets, outdirname, nside, pixlist, obscon="DARK",
         If ``True`` then log target and file information.
     scnd : :class:`bool`, defaults to ``False``
         If ``True`` then this is a ledger of secondary targets.
+    timestamp : :class:`str`, optional
+        A timestamp to use in place of that assigned by `make_mtl`.
+    exemptmf : :class:`bool`, optional, defaults to ``False``
+        If ``True`` then exempt any target that has a `TARGET_STATE`
+        driven by `MWS_FAINT_*` classes from accepting `timestamp`. These
+        targets will instead revert to a TIMESTAMP corresponding to when
+        the code was run. This is to fix a bug where "MWS_FAINT" targets
+        were not included in the (1.0.0) Main Survey target files.
 
     Returns
     -------
     Nothing, but writes the `targets` out to `outdirname` split across
     each HEALPixel in `pixlist`.
     """
     t0 = time()
+    # ADM a dictionary to hold header keywords for the ouput file.
+    hdr = {}
 
     # ADM in case an integer was passed.
     pixlist = np.atleast_1d(pixlist)
 
     # ADM execute MTL.
     mtl = make_mtl(targets, obscon, trimcols=True)
 
+    # ADM if requested, substitute a bespoke timestamp.
+    if timestamp is not None:
+        hdr["TIMESTAMP"] = timestamp
+        hdr["EXEMPTMF"] = exemptmf
+        # ADM check the timestamp is valid.
+        _ = check_timestamp(timestamp)
+        origts = mtl["TIMESTAMP"].copy()
+        mtl["TIMESTAMP"] = timestamp
+        # ADM don't use the bespoke timestamp for MWS_FAINT targets.
+        if exemptmf:
+            ii = np.array(["MWS_FAINT" in ts for ts in mtl["TARGET_STATE"]])
+            mtl["TIMESTAMP"][ii] = origts[ii]
```

Comment:
It appears that this sets the original (i.e. later, current) timestamp for any target with MWS_FAINT set, when I think we want this only for targets that have *only* MWS_FAINT set, i.e. could not have appeared in a bright mtl due to another bit being set.  In the test healpixels I haven't found a case where that happens, but there are cases where they also pass LRG cuts (e.g. TARGETID 39627601496506987 in mtl-bright-hp-6466.ecsv), and I think in principle this could happen for bright bits as well.  Please check.

Reply:
This doesn't set the current timestamp for any target with `MWS_FAINT` set. It only sets the current timestamp for targets _for which `MWS_FAINT` is driving the priority_. As `MWS_FAINT` is at the lowest priority for first-pass targets, this correctly only finds objects that have _only_ `MWS_FAINT` set.

It's a bit of trickery, admittedly, but I originally spent an hour tracking down all of the corner-cases where bits related to `MWS_FAINT_*` were set but no other bits, and no secondary matches, and only in bright time. This became very complicated. I'm far more confident that targets for which `MWS_FAINT` is "in control" are the targets that are new. And if `MWS_FAINT_*` _weren't_ the lowest-priority targets, we wouldn't want to add them anyway, as they'd bump other targets.

I think we can be confident that this was a reasonable choice, given the extensive vetting of the new ledgers by you, me and @ashleyjross this morning.

I've added a longer comment explaining this snippet of code.